README.md
LaserTank
=========

This is a mirror of the source code of the computer puzzle game LaserTank by Jim Kindley.

LaserTank was originally created by Jim in 1995. Starting with version 3.1 in 2000, Jim released the game and source code into the public domain.

I have downloaded the different source code releases from the LaserTank website on Internet Archive's Wayback Machine, and combined them into this Git repository.

In addition, I have compiled Jim's different changelogs from this repo and the website into the `CHANGELOG` file, and added a `LICENSE` file to formalize its release into the public domain.

For reference, the license boilerplates Jim Kindley added to the source code are as follows:

```
/*******************************************************
 **             LaserTank ver 4.0                     **
 **               By Jim Kindley                      **
 **               (c) 2001                            **
 **         The Program and Source is Public Domain   **
 *******************************************************
```

```
(*******************************************************
 **             Laser Tank ver 3.1.8                  **
 **               By Jim Kindley                      **
 **               (c) 1999,1998,1997,1996             **
 **         The Program and Source is Public Domain   **
 *******************************************************)
```


CHANGELOG


* modifications since LaserTank 3.1
# Bug Fixes                                      
============================

11-18-00

*Rewrote the program in C using the free lcc-win32 compiler. This can be downloaded
at http://www.cs.virginia.edu/~lcc-win32/ it is pretty cool.
The new program is a little faster and a little bigger, and is a true 32 bit program.

*Graphics have been improved:
	User loadable graphics by placing a game.bmp & mask.bmp file
	in the working directory. 

Made the mirrors and anti-tanks transparent so you can see beneath them. This
	means if you push one on a sunken block or tank mover you will see the sunken
	block or tank mover under it.

	Anti-tanks fire red now.

	System now handle more than 16 colors.

*Added More sounds.

*Added Keyboard buffer:
	The program now store moves & shots as fast as you enter them. 
	This was a cool byproduct of the next two items.

*Mouse Support:
	Left click to move and right click to shoot. The system can buffer up 
	to 20 clicks and calculates the movements during processing. 

*F4 - Replay. Try it ! Play a level 10 moves and them hit F4. This uses the recording buffer
	which is always on now. It still wont save the data unless RECORDING is on.
	P.S. The keyboard buffer is the recording buffer.

*RestartUndo, will now work even if you are recording.

11-19-00

# High Score system is back on line
	The last version didn't record high scores & didn't change HS files when LVL files
	changed.

*Added Direct Entry & Difficulty filter to "Load Level"
	You can now directly enter the level number falowed by [Enter] in the Load Level
	Dialog. If you press [Search] from the "Load Level" Dialog, There is a new Option
	to filter levels by difficulty.

Note: Debug is still turned on. ( The variable names on the right of the window.)

11-25-00

* Added Tunnels.
	You can have up to 32 pair of tunnels / transporters. If a tunnel has no match
	then it is a tunnel to HELL ( the tank will Die, all objects will vanish). 

# many editor relaited problems are fixed ( from the migration to C from pascal)

# Mouse Move routine Updated.
	Mouse Move routine can go to any valid location. The path will not go over 
	anthing but ground ( or a bridge ). If you click on a flag, tank mover or tunnel,
	the tank will move there.
# Update ALL the file routins to be more Windows 95 ( i.e. 32 type) routines. I don't know
	what advantage this will have, but we will see.

12-01-00

* Added Editor Rotating Objects
	Hold the shift key whyle clicking on a Mirror, Rotory Mirror, Tank Mover or 
	Anti-Tank in the editor mode and the object will rotate 90 Deg.

# Changed program so configuration data is stored in Ltank32.ini ( not the old Ltank.ini)

# More Editor Related Problems fixed; Starting location & Tunnel ID messup

# Tunnels will now move an object that was blocked when the blocking object is moved.
	This also works with the Tank.
( Some day I will have to write a paper on tunnels when there are more that 2 with
  the same ID #. You can have as many as you want but the logic gets real wierd,
  especially when you start blocking the tunnels )

* Added "Change Graphics" Option. The program can now use the internal graphics, Load
	External game.bmp & mask.bmp files or Load Graphic File Sets. The preferences
	are saved in the ini file.

12-8-00

* added Ctrl+V, Ctrl+C & Ctrl+G as Accelerators ( modified proograme to use acc table)

# Removed key hold down repeat for movement & spacebar.

* Added [Backspace]
	Backspace is a 10 level stack that will remember the last 10 levels played. It is
	reset whenever the program is started or a Level file is changed.

* Added [P]revious
	This is the same as skip but in the other direction. It is affected by the 
	Difficulty settings and the "Skip Complete" feature.

* Added Shift feature to the editor
	The play field can be shifted in all 4 directions.

* Added "Save As" in the editor & Modified "Save".
	"Save As" is the old "Save" routine, it asks for a file name and a level number.
	"Save" will now save the edited level to the current file & level Number.

* Added Grid Lables, this will help in discusing levels.

12-15-00

# Modified Tunnels
	The Maximum number of tunnel ID's was reduced to 8. The tunnels still operate
	like they did before. They now have a color band to indicate which tunnels are
	connected.

# Modified what happens when you hold down a key. Now the key will repeat only when the tank
	or a laser shot is not doing anything. This is the best of all worlds. During long
	laser shots you can press a direction key as many times as you like to preload the
	buffer, but if you hold down a direction key, only 1 key will be entered.If you want 
	to move a long distance just hold down the key like before, just release 1 ( or 2)
	blocks away.

# Fixed:	Wrong tankmover graphic under a moving tank.
		Saving from the editor w/ a new level would write over level 1.
		pressing letters now works in the Difficulty dialog.
		AutoRecord Check mark works properly.
		Undo after getting shot by first turn error fixed.

# Modified Opening Graphic < Note Ver 4.0 > although we will keep it 3.2.x until all
	BETA testing is complete.

12-22-00
# fixed Local HS part of Global High Score Dialog

# changed Editor so you can exit without saving. This got mixed up in the last Fix.

# Fixed a few Playback Bugs

# You can now stop a playback before you get to the end of the file and continue
	the Game. This will make it easier to tweek a playback file, just remember to
	turn on recording if you want to save it.

# Conveyors going into tunnels now works. I strainge side effect of this is, if you
	enter a tunnel and the exit side is blocked on the sides by conveyors pointing
	at you, you can reenter the tunnel by bumping into the conveyor.

* Added Completion Summary to "You Win" window

1-9-01
* Added Ice & Thin Ice Object's
	Ice will slide anything until off of the ice.
	Thin Ice only works once, then it turn into water.


LTANK.H
/*******************************************************
 **             LaserTank ver 4.0                     **
 **               By Jim Kindley                      **
 **               (c) 2001                            **
 **         The Program and Source is Public Domain   **
 *******************************************************
 **       Release version 2002 by Yves Maingoy        **
 **               ymaingoy@free.fr                    **
 *******************************************************/

 #include <windows.h>

/*

Obj			BitM	Description
---     	----    -----------

0 			1		Dirt
1 			2		Tank Up ( Primary )
-			3		Tank Right
-			4		Tank Down
-			5		Tank left
2			6		Base ( Primary)
-			7		Base Alt ( animation )
-			8		Base Alt2
3			9		Water ( Primary )
-			10		Water Alt
-			11		Water Alt 2
-       	12  	Blown Up Anti Tank (down)
4			13		Solid Block NOT movable
5			14		Movable Block
6			15		Bricks
7			16		Anti-Tank Gun Up
-			17		Anti-Tank Alt
-			18		Anti-Tank Alt 2
-			19	    Movable block after pushing into water
11			20		Mirror Up-Lt
12			21		Mirror Up-Rt
13			22		Mirror Dn-Rt
14			23		Mirror Dn-Lt
15			24		One Way Up (Primary)
-			25		One Way Up Alt
-			26		One Way Up Alt2
16			27		One Way Rt (Primary)
-			28		One Way Rt Alt
-			29		One Way Rt Alt2
17			30		One Way Dn (Primary)
-			31		One Way Dn Alt
-			32		One Way Dn Alt2
18			33		One Way Lt (Primary)
-			34		One Way Lt Alt
-			35		One Way Lt Alt2
8 	      	36      Anti-Tank facing right
- 	      	37
- 	      	38
9       	39      Anti-Tank facing down
-       	40
-       	41
10      	42      Anti-Tank facing left
-       	43
-       	44
19      	45      Crystal Block
-       	46      Crystal when hit by tank
20			47		Roto Mirror Up-Lt
21			48		Roto Mirror Up-Rt
22			49		Roto Mirror Dn-Rt
23			50		Roto Mirror Dn-Lt
-       	51      Crystal when hit by Anti-tank
-       	52      Blown Up Anti Tank (right)
-       	53      Blown Up Anti Tank (left)
-       	54      Blown Up Anti Tank (up)
24			56		Ice
25			57		Thin Ice
01dddddx	55		Worm Hole / Tunnel

*/

// Language Section for Lasertank
// Import this from its own file

#include "lt32l_us.h"

// Game Defaults
#define LevelData       "LaserTank.lvl"    // Default Level Data File
#define INIFileName     "LaserTank.ini"
// MAX text size for a lang_text
#define MAX_LANG_SIZE    300



#define MaxObjects	    26              // Maximum number of game objects used
#define MaxBitMaps      58             	// the maximum number of BitMaps ( + 1 )
#define XOffset         17             	// Game Board Offset from Left
#define YOffset         17             	// Game Board Offset from top
#define ani_delay       4              	// animation Delay in GameDelay units
#define GameDelay       50             	// Main Operation Loop
#define WM_GameOver     WM_USER +1     	// Send Message if there are no more levels
#define WM_Dead         WM_USER +2     	// Send Message when you die
#define WM_NewHS	    WM_USER +3      // Send to Display New High Score Dialog
#define WM_SaveRec      WM_USER +4      // Send to Save Recording
#define App_Class		"LaserTC2"		// The Class Name
#define UndoBufStep		200				// Groth Amount of Undo Buffer
#define UndoMax			10000			// Max Amount of Undo Buffer
#define RecBufStep		10000			// Groth amount of Rec Buffer
#define RecMax			65500			// Max recording steps saved to file
#define SlowPBSet		5				// Delay of Slow Playback setting
#define MaxMBuffer		20				// Size of Mouse Buffer
#define Obj_Water		3				// Object Number 3 is water
#define Obj_Ice			24				// Ice Object
#define Obj_ThinIce		25				// Object Number of Thin Ice
#define Obj_Tunnel		0x40			// Object 01dddddX = Tunnel
#define LTG_ID			"LTG1"			// ID field of LTG file
#define psRLLOn			"RLL"          // Profile String for remember last level
#define psRLLN			"RLLFilename"  // remember last level file name
#define psRLLL			"RLLLevel"     // Level number
#define psAni			"Animation"
#define psSound			"Sound"
#define psSize			"Size"			// 1 = small, 3 = large
#define psXpos			"PosX"
#define psYpos			"PosY"
#define psUser 			"Player"
#define psPBA 			"Record Author"
#define psARec			"Auto_Record"
#define psSCL			"SkipComLev"
#define psDiff			"Diff_Setting"
#define psGM			"Graphics_Mode"
#define psGFN			"Graphics_File"
#define psGDN			"Graphics_Dir"
#define psDW			"DisableWarnings"
#define psYes			"Yes"
#define psNET			"NETWORK_INI"	// 1 = Use INI file from C:\Windows
#define BADMOVE 		256

typedef char TPLAYFIELD [16][16]; 		// Matrix of G.O. types

typedef struct tLevel               // Level Data from File
{
  TPLAYFIELD PF;               // Object Grid
  char LName[31];  			// Level Name
  char Hint[256];  			// Hint for this level
  char Author[31]; 			// the Author of the Level
  WORD SDiff; 				// Score Difficulty
}TLEVEL;

typedef struct tRecordRec		// Recording Header
{
	char LName[31];				// Level Name
	char Author[31];			// Author of the recording
	WORD Level;					// Level Number
	WORD Size;					// Size of Data -- Data to fallow
}TRECORDREC;

typedef struct tSearch			// Search Record used in Level Load
{
	int mode,					// 1=title, 2 = author
		SkipComp;				// True = Skip Completed
	WORD Diff;					// Difficulty charecter
	char data[61];				// Search String
}TSEARCH;

typedef struct tTankRec              // Store the Tank & Laser information
{
  int X, Y, Dir, Firing, Good;		// Good is used for Tunnel Wait in Game.Tank
}TTANKREC;

typedef struct tGameRec
{
  TPLAYFIELD	PF,               // Store Game Objects
  				PF2,              // Store Objects Under Stuff ( Ground, conveyor)
  				BMF,              // Bitmaps for Objects
  				BMF2;             // Bitmaps for Under Stuff ( Bridges )
  WORD 			ScoreMove,        // Move Counter
  				ScoreShot;        // Shot Counter
  DWORD			RecP;			  // Recording Pointer
  TTANKREC		Tank;             // Tank Data
}TGAMEREC, *PGAMEREC;

typedef struct tXYRec			// Use in BMA
{
	int X,Y;					// X & Y Location in Big Butmap
}TXYREC;

typedef struct tXYZRec			// Use for Mouse Buffer
{
	int X,Y,Z;
}TXYZREC;

typedef struct tHSRec					// High Score Record
{
	WORD moves, shots;
	char name[6];						// Initials
}THSREC;

typedef struct tLTGRec			// Record for LTG graphics
{
	char Name[40];		// Name of Graphic Set
	char Author[30];	// Author of Graphics set
	char Info[245];		// 3 line Description of Graphics Set
	char ID[5];			// LTG ID = "LTG1"+0
	DWORD MaskOffset;	// Offset from the biggining of file to Mask Bitmap
}TLTGREC;

typedef struct tIceRec	// Record used for sliding on the Ice
{
	int x,y,			// Last XY position of object to move
		dx,dy,			// Direction to move in Delta Cords
		s;				// True if Sliding

}TICEREC;

/// ------------ MGY -----------
#define MAX_TICEMEM 16
typedef struct tIceMem	// Record used for sliding objects on the Ice
{
	TICEREC Objects[MAX_TICEMEM];  // MGY - mem up 16 sliding objects
	int count;          // number of current sliding objects
}TICEMEM;
extern TICEMEM SlideMem;
/// ------------ MGY -----------

// Extern's -- Complete Program Defined Global Variables
extern HINSTANCE hInst;					// Globally Defined Instance
extern char *RecBuffer;
extern int GFXOn, Game_On, Sound_On, Ani_On, RLL, ConvMoving, CurLevel,TankDirty, DWarn;
extern int AniCount, CurSelBM_L, CurSelBM_R, RB_TOS, PBHold, SpBm_Width,SpBm_Height,ContXPos;
extern int ARecord, Difficulty,SkipCL, Recording,PlayBack, EditBMWidth, Speed, SlowPB, RecBufSize;
extern HWND MainH, Ed1, Ed2, BT1, BT2, BT3, BT4, BT5, BT6, BT7,BT8, BT9;
extern TGAMEREC Game,BackUp,SaveGame;
extern TLEVEL CurRecData;
extern HBRUSH LaserColor, LaserColorG, LaserColorR;
extern HDC gDC;                      // Use this game dc for all ops
extern char FileName[], HFileName[], GHFileName[], PBFileName[], GraphFN[], GraphDN[],INIFile[];
extern HMENU MMenu;
extern int Modified,OKtoHS,VHSOn;
extern TRECORDREC PBRec, PBSRec;
extern char HelpFile[];
extern PGAMEREC UndoBuffer;
extern int UndoP,PBOpen;
extern THSREC HS;
extern HDC BuffDC;
extern HBITMAP BuffBMH;
extern TXYREC BMA[MaxBitMaps+1]; 			// Bit Map Array
extern TXYZREC MBuffer[MaxMBuffer];		// Mouse Buffer
extern int MB_TOS, MB_SP;				// TOS = Top of Stack ; SP = Stack Pointer
extern DWORD DEBUG_Time;
extern DWORD DEBUG_Frames;
extern HANDLE F1,F2,F3;					// File Handles
extern DWORD BytesMoved;				// use in ReadFile & WriteFile
extern int GraphM, FindTank;
extern TICEREC SlideO,SlideT;

// Variables for Language
extern char LANGText[SIZE_ALL][MAX_LANG_SIZE]; // All lines of Language.dat
extern char LANGFile[MAX_PATH];                // the dynamic file name for Language.dat
extern TCHAR szFilterOFN[MAX_PATH];
extern TCHAR szFilterPBfn[MAX_PATH];

// Global Function Prototypes

void FileError();
void GFXInit();									// Load All the Bitmaps
void GFXKill(void);								// Release All the Bitmaps
void PutLevel(void);							// Draw Game Board
void GameOn(int);								// Turns on/off game timer
void JK3dFrame(HDC, int, int, int, int, int);	// Draw 3D frame
void JKSelFrame(HDC, int, int, int, int, int);	// Draw Red/Green frame
void LoadLastLevel();
BOOL LoadNextLevel( int, int);
void CheckHighScore(void);						// Compare score to High Score
void Animate(void);								// Animate all Animated Objects
void MoveTank( int );
int  CheckLoc(int, int);
void MoveLaser();
void FireLaser(int, int, int, int);
void AntiTank();								// Target Anti-Tanks at Tank
char GetOBM(char);								// Translate Bitmap to Object
void BuildBMField();							// Convert Objects to bitmaps
void PutSprite (char, int, int);				// Paint BM at X,Y
void UpDateSprite(int, int);					// Repaint bitmap at x,y
void UpDateTank();								// Repaint Tank
void PutSelectors();							// Editor - paint objects to edit
void ChangeGO(int, int, int);					// Change Game Object
void SetGameSize(int);
void SavePBFile();
void InitBuffers();
void KillBuffers();
void AssignHSFile();
void UpdateUndo();
void UndoStep();
void ResetUndoBuffer();
int MouseOperation(int);						// Process Mouse Input
void ShowTunnelID();
void ConvMoveTank(int, int, int);
void IceMoveT();					// Move an tank on the Ice
void IceMoveO();					// Move an Object on the Ice
void AddKBuff(char);

// Function Prototypes for Language
void ChangeMenuText(HMENU, char*);
void InitLanguage(HMENU, HMENU);
void ConvertTabChar(char*);
HBITMAP LoadImageFile(HINSTANCE, char *, char*);
void LoadWindowText(HWND, int);
void LoadWindowCaption(HWND, int);


/* Lets try some Macros */
#define GetTunnelID(x,y) ((Game.PF[x][y] & 0x0F)  >> 1)			// 0 - 7

// mgy 18-05-2003 Tunnel is limited to 0-7.
#define GetTunnelOldID(x,y) ((Game.PF[x][y] & 0x0F)  >> 1)			// 0 - 7
//#define GetTunnelOldID(x,y) ((Game.PF[x][y] & 0x3F)  >> 1)		// 0 - 32

#define ISTunnel(x,y) ((Game.PF[x][y] & Obj_Tunnel) == Obj_Tunnel)
#define GameInProg Game.RecP && (Game.PF[Game.Tank.X][Game.Tank.Y] != 2) && (!DWarn)

